project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/organisations/RorApi.java

```diff
@@ -0,0 +1,252 @@
+        if (e.getHttpStatusCode().isPresent()) {
+          errorMessage += "Failed with http status code " + e.getHttpStatusCode().orElseThrow();
+        }
+        log.error(errorMessage, e);
+        return Optional.empty();
+      }
+
+      Optional<RorEntry> rorEntry = result.optionalBody()
+          .flatMap(this::parseJson);
+
+      if (rorEntry.isEmpty()) {
+        log.warn(
+            "No organisation with identifier %s was found on %s.".formatted(rorId, request.uri()));
+      }
+
+      return rorEntry;
+    }
+
+    private Optional<RorEntry> parseJson(String json) {
+      try {
+        RorEntry rorEntry = new ObjectMapper().configure(
+                DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+            .readValue(json, RorEntryV2.class);
+        return Optional.of(rorEntry);
+      } catch (JsonProcessingException e) {
+        log.error("Could not parse response from ROR.", e);
+        return Optional.empty();
+      }
+    }
+
+
+    /**
+     * Needs to be handled
+     */
+    private static class RorRequestException extends Exception {
+
+      private final Integer httpStatusCode;
+
+
+      public RorRequestException(Throwable cause) {
+        super(cause);
+        httpStatusCode = null;
+      }
+
+      public RorRequestException(String message, int httpStatusCode) {
+        super(message);
+        this.httpStatusCode = httpStatusCode;
+      }
+
+      public Optional<Integer> getHttpStatusCode() {
+        return Optional.ofNullable(httpStatusCode);
+      }
+    }
+
+    private record RorResponse(String body) {
+
+      public Optional<String> optionalBody() {
+        return Optional.ofNullable(body);
+      }
+
+      @Override
+      public String body() {
+        if (Objects.isNull(body)) {
+          throw new NoSuchElementException(
+              "No body is present. Please use RorResponse#optionalBody instead");
+        }
+        return body;
+      }
+
+      static RorResponse empty() {
+        return new RorResponse(null);
+      }
+
+      static RorResponse of(String body) {
+        Objects.requireNonNull(body);
+        return new RorResponse(body);
+      }
+    }
+
+    private RorResponse sendRequest(HttpRequest request) throws RorRequestException {
+      HttpResponse<String> result;
+      try (HttpClient client = HttpClient.newBuilder().version(Version.HTTP_2)
+          .followRedirects(Redirect.NORMAL).connectTimeout(
+              Duration.ofSeconds(10)).build();) {
+
+        result = client.send(request, BodyHandlers.ofString());
+        //If a valid RoRId was provided but the ID does not exist we fail
+      } catch (IOException e) {
+        throw new RorRequestException(e);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new RorRequestException(e);
+      }
+
+      if (result.statusCode() == HttpStatus.NOT_FOUND_404) {
```

Comment:
int in comparison vs an enum?

Reply:
That is not an enum but a static int value 😄